File: Python/面试题08.03.py
class Solution:
    def findMagicIndex(self, nums: List[int]) -> int:
        # for i, num in enumerate(nums):
        #     if i == num:
        #         return i
        # return -1

        # i, j = 0, len(nums) - 1
        # while i < j:
        #     m = (i + j) // 2
        #     if nums[m] >= m:
        #         j = m
        #     else:
        #         i = m + 1
        # ans = i if nums[i] == i else -1 #解决不了元素带有重复的问题
        # return ans

        i, j = 0, len(nums)
        while i < j:
            if i == nums[i]:
                return i
            elif i < nums[i]:
                i = nums[i]
            else:
                i += 1
        return -1

File: C++/面试题 08.03.cpp
class Solution {
public:
    int findMagicIndex(vector<int>& nums) {
        int min_ = -1;
        // for (vector<int>::iterator it = nums.begin(); it < nums.end(); ++it)
        for (int i = 0; i < nums.size(); ++i)
        {
            if (nums[i] == i)
            {
                if (min_ == -1 or i < min_) #居然写了个or？ ||
                {
                    min_ = i;
                }
            }
        }
        return min_;
    }
};
